src/g3log/logmessage.hpp

```diff
@@ -87,7 +88,8 @@ namespace g3 {
       // Complete access to the raw data in case the helper functions above
       // are not enough.
       //
-      std::time_t _timestamp;
+      // TODO: unique_ptr should be considered instead of shared_ptr
+      std::shared_ptr<Timestamp> _timestamp;
```

Comment:
I would then advice on the function pointer method I suggested earlier. With the difference that it would return your time struct which points to the time of the call into the function. 

I.e I envision 

1) the timestamp provider class to not be needed.
2) the time struct can be stored inside the LogMessage struct
3) the LogMessage struct would provide with similar API as today a way of formatting the time into a string
4) possibly using the posix time struct (why reinvent the wheel?)

Does the above make sense?